### DISCUSSION TITLE
Suggestions on CANopen HW devices for testing

### DISCUSSION THREAD

**Discussion by sveinse** [canopen-python/canopen#434] [General]
Hi

In the development of [canopen-asyncio](https://github.com/sveinse/canopen-asyncio), I have want to test more aspects of CANOpen than with HW I currently have.

Does anyone have any recommendation for a well-implemented, suitable CANopen device that can be used to test the library against? Some device that is able to do 402 and LSS would be nice. Preferably in Norway or Scandinavia (but EU is ok).

**Comment by acolomb** [comment]
I've used the motor controllers from Nanotec extensively with this library. Pretty well documented and a clean CANopen implementation, which is also used as the basis for other protocols like Modbus or Ethernet. They often also have a USB interface to connect with their proprietary configuration software in various ways. The only issue I had to work around IIRC was a two second pause from entering pre-operational state before they actually start responding to SDO requests. I hope you can find something there at a decent price, maybe integrated to a drive even. 

miControl also has some decent small controllers. But I haven't used them with DS402 yet (beware, some cheaper ones don't support it). The docs are pretty good and they supply their own Python-based configuration software, although unfortunately not using this library.

From my experience, look at the device manuals first. If they contain a decent explanation of CANopen in general, then it's usually also a good implementation adhering to the standard. 

Oh, some manufacturers actually define objects with subindex != 0 but which are simple variables and not part of a record or array. Beware of those, although this library usually copes quite well with such quirks.

**Comment by ericbaril72** [comment]
While keeping in line with the **Title** of Sveinse suggestion:  would it be outside the scope of this CANopen library to include more device profiles examples ?

I have used this CANopen library for the longest time ( >6-7 years ), mainly to better understand the different data transfer modes between 2 nodes.  

My Findings :  
 - pretty much all device manufacturers only use a very small subset of Cia-CANopen functionnalities.
 - Not all device manufacturers use the same "emergency" codes for similar errors ( ex: accessing inexistant SDO idx )
 - Not all devices accept all types of transfers

Would it make sense to have example sub-directories that we could propose   mfg/PN test cases ?  or is it outside the scope ?

I have been using multiple brands of IO-modules and PLC ( mainly IFM ) and been using this library to develop projects to interface new sensors before coding it into the PLC  ( Codesys is not very friendly when it comes to unit tests, both 2.3 and 3.5, especially when have to support both )

Few examples that comes to mind:
some devices can only be configured in pre-op mode, others do not care and can be dynamicaly reconfigured

I often looked at ds402 example and cannot relate ( never used a 402 device ) to propose a ds401 example.
Based on experience, a proposed ds401 that "WORKS" with IFM CR2032 ioModule https://www.ifm.com/ca/en/product/CR2032
                                                 would not necessarily work with a CR2042 ioModule https://www.ifm.com/ca/en/product/CR2042

THUS incorporating examples PER mfg/pn  ( supplied by collaborators on the lib )
